src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,153 @@
+    private Node<T> tail;
+    private int size;
+
+    public static class Node<T> {
+        private T value;
+        private Node<T> next;
+        private Node<T> prev;
+
+        public Node(T value, Node<T> prev, Node<T> next) {
+            this.prev = prev;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(value, tail, null);
+        if (tail == null) {
+            head = newNode;
+        } else {
+            tail.next = newNode;
+        }
+        tail = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Invalid Index: "
+                    + index
+                    + ", Size: "
+                    + size);
+        }
+        if (index == size) {
+            add(value);
+            return;
+        }
+        Node<T> current = node(index);
+        Node<T> newNode = new Node<>(value, current.prev, current);
+        if (current.prev == null) {
+            head = newNode;
+        } else {
+            current.prev.next = newNode;
+        }
+        current.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return node(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = node(index);
+        T oldValue = current.value;
+        current.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = node(index);
+        if (current.prev == null) {
+            head = current.next;
+        } else {
+            current.prev.next = current.next;
+        }
+        if (current.next == null) {
+            tail = current.prev;
+        } else {
+            current.next.prev = current.prev;
+        }
+        size--;
+        return current.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        while (current != null) {
+            if ((object == null && current.value == null)
+                    || (object != null && object.equals(current.value))) {
+                if (current.prev == null) {
+                    head = current.next;
+                } else {
+                    current.prev.next = current.next;
+                }
+                if (current.next == null) {
+                    tail = current.prev;
+                } else {
+                    current.next.prev = current.prev;
+                }
+                size--;
+                return true;
```

Comment:
Be mindful of the possibility of 'null' being passed as a method parameter in the remove(T object) method. This code handles it, but make sure to consider null cases whenever you are working with objects.